### DISCUSSION TITLE
Multiple Self-hosted Github actions runners

### DISCUSSION THREAD

**Discussion by sgrmgj** [community/community#26049] [Actions]
<p>I have two self hosted github action runners - </p>
<ol>
<li>
<p>The first runner is running in my dev env</p>
</li>
<li>
<p>The second runner is running in my qa env.</p>
</li>
</ol>
<p>Is there a way to select these runners based on which branch the Java CI action is being executed.</p>
<p>so e.g. I want to run my master branch on the runner hosted in QA env and I want to run my feature branch run on the runner which is hosted in the dev env.</p>
<p>Is there a way to do this. Currently I see that I can only use _ <strong>self-hosted </strong> _label but how can I choose - </p>
<pre><code>run:
 on: my-runner
</code></pre>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/sgrmgj-quotient">@sgrmgj-quotient</a> ,</p>
<p>Currently self hosted runner supports below <a href="https://help.github.com/en/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow#about-self-hosted-runner-labels" rel="nofollow noopener">labels</a>:</p>
<ul>
<li>self-hosted: Default label applied to all self-hosted runners.</li>
<li>linux, windows, or macos: Applied depending on operating system type.</li>
<li>x86, x64, ARM, or ARM64: Applied depending on architecture type.</li>
</ul>
<p>For example, you can use below code to choose a macos self-hosted runner.</p>
<pre><code>runs-on: [self-hosted, macos]
</code></pre>
<p>If your two runners used different OS/architecture, then they can be specified separately in your workflow.</p>
<p>Github engineering team are going to support custom labels, should be completed by 31 March 2020.</p>
<p>Let’s wait for the feature to be released, then you can use custom label to specify the self-hosted runner in workflow. You could keep an eye on <a href="https://github.blog/changelog/" rel="nofollow noopener">Changelog</a> site for the new features of Github Actions. </p>
<p>Thanks!</p>

**Comment by sgrmgj** [comment]
<p>Thank you so much for the reply.</p>

**Comment by FinalDes** [comment]
<p>how to add different type of linux distribution label on custom runner?</p>

**Comment by mscoutermarsh** [comment]
<p>Custom labels are now available for self-hosted runners.</p>
<p><a href="https://github.blog/changelog/2020-04-22-github-actions-organization-level-self-hosted-runners/" rel="nofollow noopener">https://github.blog/changelog/2020-04-22-github-actions-organization-level-self-hosted-runners/</a></p>

**Comment by sfynx** [comment]
<aside class="quote no-group" data-username="sgrmgj-quotient" data-post="1" data-topic="17764">
<div class="title">
<div class="quote-controls"></div>
 sgrmgj-quotient:</div>
<blockquote>
<p>Is there a way to select these runners based on which branch the Java CI action is being executed.</p>
</blockquote>
</aside>
<p>And most importantly: how do we enforce this assignment? Code in unmerged feature branches (including the workflow file itself) is untrusted, so any labels configured there cannot be relied upon.</p>

**Comment by ogind-bdhavale** [comment]
You can use GitHub Actions matrix strategies to define a matrix of different job configurations. For e.g. you can create matrix of runner and git branch to use.
```
jobs:
  example_matrix:
    strategy:
      matrix:
        os: [QA_runner, Dev_runner]
        version: [feature_branch, master_branch]
```

https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs
